clippy_lints/src/methods/manual_as_ref_deref.rs

```diff
@@ -20,14 +27,23 @@ pub(super) fn check(
     is_mut: bool,
     msrv: &Msrv,
 ) {
-    if !msrv.meets(msrvs::OPTION_AS_DEREF) {
-        return;
-    }
-
     let same_mutability = |m| (is_mut && m == &hir::Mutability::Mut) || (!is_mut && m == &hir::Mutability::Not);
 
-    let option_ty = cx.typeck_results().expr_ty(as_ref_recv);
-    if !is_type_diagnostic_item(cx, option_ty, sym::Option) {
+    let target = 'target: {
+        let target_ty = cx.typeck_results().expr_ty(as_ref_recv);
+        if is_type_diagnostic_item(cx, target_ty, sym::Option) {
+            break 'target Target::Option;
+        }
+        if is_type_diagnostic_item(cx, target_ty, sym::Result) {
+            break 'target Target::Result;
+        }
+        return;
+    };
+
+    if target == Target::Option && !msrv.meets(msrvs::OPTION_AS_DEREF) {
+        return;
+    }
+    if target == Target::Result && !msrv.meets(msrvs::RESULT_AS_DEREF) {
         return;
     }
```

Comment:
Using `get_type_diagnostic_name()` would let you simplify this part:

```rust
    let target_ty = cx.typeck_results().expr_ty(as_ref_recv);
    let target = match get_type_diagnostic_name(cx, target_ty) {
        Some(sym::Option) if msrv.meets(msrvs::OPTION_AS_DEREF) => Target::Option,
        Some(sym::Result) if msrv.meets(msrvs::RESULT_AS_DEREF) =>Target::Result,
        _ => return,
    };
```

Reply:
Thanks. This is really nice. I've applied those suggestions.